### PR TITLE
Remove stream key from log statements

### PR DESCRIPTION
**Pull request by mv2731**
Previously, both the ffmpeg and OBS log exposed the stream key in the HLS ingestion URL.
This pull request fixes the issue by doing the following:

1. Set a field called printable in the ffmpeg_muxer struct, the printable equivalent of path
2. Set a member of main_params to be print_file, the printable equivalent of file 
3. Edit the code logic surrounding sensitive prints, so that the printable version is checked first, and if set, printed. 
4. Write a logging callback function for ffmpeg and replace the stream key with "{stream_key}" if ever it is detected

Relevant issues:
https://github.com/obsproject/obs-studio/pull/3114

### CODE REVIEW

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 1]
```diff
@@ -27,8 +27,16 @@
 #include <stdlib.h>
 #include "ffmpeg-mux.h"
 
+#include <obs-module.h>
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Is obs-module.h needed?

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 2]
```diff
@@ -70,6 +80,7 @@ static inline void resize_buf_free(struct resize_buf *rb)
 
 struct main_params {
 	char *file;
+	char *print_file;
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: printable_file and add a comment about this field.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 3]
```diff
@@ -45,6 +45,7 @@ struct ffmpeg_muxer {
 	int64_t stop_ts;
 	uint64_t total_bytes;
 	struct dstr path;
+	const char *printable_path;
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: Also add a comment about printable_path: It's the path with stream key information removed.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 4]
```diff
@@ -420,13 +443,16 @@ static bool ffmpeg_hls_mux_start(void *data)
...
+	stream->printable_path = path_str;
+	info("Writing to path '%s'...", stream->printable_path);
 
 	return true;
 }
 
 static int deactivate(struct ffmpeg_muxer *stream, int code)
 {
+	obs_service_t *service;
+	const char *temp;
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: These two variables are not needed anymore.

Comment (mv2731) [reply plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: yep thanks!

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 5]
```diff
@@ -257,6 +302,16 @@ static bool init_params(int *argc, char ***argv, struct main_params *params,
 
 	*p_audio = audio;
 
+	get_opt_str(argc, argv, &global_stream_key, "stream key");
+	if (strcmp(global_stream_key, "") != 0) {
+		struct dstr tmp = {0};
+		dstr_copy(&tmp, params->file);
+		dstr_replace(&tmp, global_stream_key, "{stream_key}");
+		strcpy(params->file, tmp.array);
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Shouldn't tmp.array be assigned to printable_file? I don't think HLS streaming is working with this change.

Comment (mv2731) [reply plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Yes! I think this occurred when I changed the names--you're completely right. Even though the log looks fine it doesn't successfully stream to YouTube. Thanks for the catch.

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 6]
```diff
@@ -70,6 +79,8 @@ static inline void resize_buf_free(struct resize_buf *rb)
 
 struct main_params {
 	char *file;
+	/* printable_file is file with any stream key information removed */
+	char printable_file[4096];
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Can struct dstr be used for printable_file?

**Review by ushadow — COMMENTED** [review]
Please see comment.

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 7]
```diff
@@ -257,6 +304,16 @@ static bool init_params(int *argc, char ***argv, struct main_params *params,
 
 	*p_audio = audio;
 
+	get_opt_str(argc, argv, &global_stream_key, "stream key");
+	if (strcmp(global_stream_key, "") != 0) {
+		struct dstr tmp = {0};
+		dstr_copy(&tmp, params->file);
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: You can just directly copy to printable_file. Don't need tmp.

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 8]
```diff
@@ -471,8 +528,11 @@ static inline int open_output_file(struct ffmpeg_mux *ffm)
 		ret = avio_open(&ffm->output->pb, ffm->params.file,
 				AVIO_FLAG_WRITE);
 		if (ret < 0) {
-			fprintf(stderr, "Couldn't open '%s', %s",
-				ffm->params.file, av_err2str(ret));
+			fprintf(stderr, "Couldn't open '%s', %s\n",
+				ffm->params.printable_file.len
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: dstr_is_empty(&ffm->params.pritable_file), and same elsewhere.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 9]
```diff
@@ -217,6 +219,27 @@ static void log_muxer_params(struct ffmpeg_muxer *stream, const char *settings)
...
+		return NULL;
+	stream_key = obs_service_get_key(service);
+	return stream_key;
+}
+
+static void add_stream_key(struct dstr *cmd, struct ffmpeg_muxer *stream)
+{
+	const char *stream_key;
+
+	stream_key = get_stream_key(stream);
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: I think it's not necessary to pass stream key in the command when it is not HLS. We can store stream_key as a member field in struct ffmpeg_muxer and set only in ffmpeg_hls_mux_start.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 10]
```diff
@@ -45,6 +45,8 @@ struct ffmpeg_muxer {
 	int64_t stop_ts;
 	uint64_t total_bytes;
 	struct dstr path;
+	/* printable_path is path with any stream key information removed */
+	const char *printable_path;
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: Use struct dstr for printable_path, then you don't need to worry about if the array pointed by the pointer could be deallocated or not.

**Review by ushadow — CHANGES_REQUESTED** [review]
Please see the comments.

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 11]
```diff
@@ -372,7 +386,7 @@ static bool ffmpeg_hls_mux_start(void *data)
 	struct ffmpeg_muxer *stream = data;
 	obs_service_t *service;
 	const char *path_str;
-	const char *stream_key;
```
Comment (ushadow) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c]: Just use stream_key.

**Review by ushadow — CHANGES_REQUESTED** [review]

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 12]
```diff
@@ -200,6 +213,40 @@ static bool get_audio_params(struct audio_params *audio, int *argc,
...
+
+	switch (level) {
+	case AV_LOG_INFO:
+		fprintf(stdout, "info: [ffmpeg_muxer] %s", out.array);
+		fflush(stdout);
+		break;
+
+	case AV_LOG_WARNING:
+		fprintf(stdout, "warning: %s[ffmpeg_muxer]%s %s%s%s",
+			ANSI_COLOR_MAGENTA, ANSI_COLOR_RESET, ANSI_COLOR_YELLOW,
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Just use magenta for the whole line including warning.

File: plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c [hunk 13]
```diff
@@ -200,6 +213,40 @@ static bool get_audio_params(struct audio_params *audio, int *argc,
...
+			out.array, ANSI_COLOR_RESET);
+		fflush(stdout);
+		break;
+
+	case AV_LOG_ERROR:
+		fprintf(stderr, "error: %s[ffmpeg_muxer]%s %s%s%s",
+			ANSI_COLOR_BLK, ANSI_COLOR_RESET, ANSI_COLOR_RED,
```
Comment (ushadow) [inline plugins/obs-ffmpeg/ffmpeg-mux/ffmpeg-mux.c]: Just use red for the whole line including error.

**Review by ushadow — CHANGES_REQUESTED** [review]
Please see comments.

**Pull request merged**